Split deployment gates setup into JIT and preconfigured pages (#37364)
DataDog/documentation: content/en/deployment_gates/setup/preconfigured.md

---
title: Set Up Preconfigured Deployment Gates

> description: "Create gates and rules in Datadog ahead of time, then reference them by service and environment at deployment time."

Reviewer: ```suggestion
description: "Create gates and rules in Datadog in advance, then reference them by service and environment at deployment time."
```

further_reading:
- link: "/deployment_gates/setup/jit"
  tag: "Documentation"

>   text: "Set up JIT Deployment Gates"

Reviewer: ```suggestion
  text: "Set up Just-In-Time (JIT) Deployment Gates"
```

Deployment Gates are in Preview. If you're interested in this feature, complete the form to request access.
{{< /callout >}}

With **preconfigured** Deployment Gates, gates and rules are persisted in Datadog and referenced by service and environment at evaluation time. Preconfigured gates are a good fit when you want to share rules across many deployments, manage configuration in Terraform, or let non-CI users edit rules in the Datadog UI.

> Looking to define rules inline in your deployment config? See [JIT Deployment Gates][5].

Reviewer: ```suggestion
Looking to define rules inline in your deployment config? See [Just-In-Time (JIT) Deployment Gates][5].
```

2. Click {{< ui >}}Create Gate{{< /ui >}}.
3. Configure the following settings:
   - {{< ui >}}Service{{< /ui >}}: The service name (example: `transaction-backend`).
   - {{< ui >}}Environment{{< /ui >}}: The target environment (example: `dev`).
   - {{< ui >}}Identifier{{< /ui >}} (optional, default value is `default`): Unique name for multiple gates on the same service/environment. Use this to:
     - Allow different deployment strategies (example: `fast-deploy` vs `default`)
     - Distinguish deployment phases (example: `pre-deploy` vs `post-deploy`)
     - Define canary stages (example: `pre-deploy` vs `canary-20pct`)
   - {{< ui >}}Evaluation Mode{{< /ui >}}: Enable {{< ui >}}Dry Run{{< /ui >}} to test gate behavior without impacting deployments. The evaluation of a dry-run gate always responds with a pass status, but the in-app result reflects the real evaluation. This is useful when performing an initial evaluation of the gate behavior without impacting the deployment pipeline.

## Add rules to a gate

Each gate requires one or more rules to evaluate. All rules must pass for the gate to succeed. For each rule, specify:

> 1. {{< ui >}}Name{{< /ui >}}: A descriptive label that shows up on the [Deployment Gates Evaluations][7] page (for example, `Check all P0 monitors`).

Reviewer: ```suggestion
1. {{< ui >}}Name{{< /ui >}}: A descriptive label that appears on the [Deployment Gates Evaluations][7] page (for example, `Check all P0 monitors`).
```

2. {{< ui >}}Type{{< /ui >}}: Select {{< ui >}}Monitor{{< /ui >}} or {{< ui >}}Faulty Deployment Detection{{< /ui >}}.
3. Additional settings based on the selected rule type. See [Rule types](#rule-types) for the available options.
4. {{< ui >}}Evaluation Mode{{< /ui >}}: When a rule is set as a {{< ui >}}Dry Run{{< /ui >}}, its result is not taken into account when computing the overall gate result.

## Rule types

For the full schema and all available options, see the [Deployment Gates API reference][4].

{{< tabs >}}

> {{% tab "Monitors" %}}

Reviewer: ```suggestion
{{% tab "Monitor" %}}
```

{{% /tab %}}
{{< /tabs >}}

> ## Evaluate a gate from your pipeline

Reviewer 2: Not from this PR, but we can tackle it now: I'd add identifier: default to the examples here. A lot of customers add identifier to their gates because they don't know what it is, and then cannot evaluate the gate because they miss it when copy-pasting the example

Author: Done!

After the gate is configured, request an evaluation when deploying the related service, and decide whether to block or continue the deployment based on the result.

{{< tabs >}}
{{% tab "datadog-ci CLI" %}}
The [datadog-ci][1] `deployment gate` command runs the evaluation in a single command:

```bash
datadog-ci deployment gate --service transaction-backend --env staging
```

If the Deployment Gate contains APM Faulty Deployment Detection rules, also specify the version (for example, `--version 1.0.1`).

The command:

- Sends a request to start the gate evaluation and blocks until the evaluation is complete.
- Provides a configurable timeout for how long to wait for an evaluation.
- Has built-in automatic retries for errors.
- Accepts `--fail-on-error` to customize behavior on unexpected Datadog errors.

The `deployment gate` command is available in datadog-ci versions v3.17.0 and above.

**Required environment variables**:

- `DD_API_KEY`: Your [API key][2].

> - `DD_APP_KEY`: Your [Application key][3].

Reviewer: ```suggestion
- `DD_APP_KEY`: Your [application key][3].
```

{{% /tab %}}
{{% tab "GitHub Actions" %}}

> The [`Datadog Deployment Gate GitHub Action`][4] runs the evaluation as part of a workflow.

Reviewer: ```suggestion
The [Datadog Deployment Gate GitHub Action][4] runs the evaluation as part of a workflow.
```

          # Your deployment commands here
```

If the Deployment Gate contains APM Faulty Deployment Detection rules, also specify the version (for example, `version: 1.0.1`).

The action:

- Sends a request to start the gate evaluation and blocks until the evaluation is complete.
- Provides a configurable timeout for how long to wait for an evaluation.
- Has built-in automatic retries for errors.
- Accepts `fail-on-error` to customize behavior on unexpected Datadog errors.

**Required environment variables**:

- `DD_API_KEY`: Your [API key][2].

> - `DD_APP_KEY`: Your [Application key][3].

Reviewer: ```suggestion
- `DD_APP_KEY`: Your [application key][3].
```

## Recommendation for first-time onboarding

> When integrating Deployment Gates into your Continuous Delivery workflow, an evaluation phase helps confirm the product is working as expected before it impacts deployments. Use the Dry Run evaluation mode and the {{< ui >}}Deployment Gates Evaluations{{< /ui >}} page:

Reviewer: Add link to Deployment Gates Evaluation page

Author: Done!

1. Create a gate for a service and set the {{< ui >}}Evaluation Mode{{< /ui >}} to {{< ui >}}Dry Run{{< /ui >}}.
2. Add the gate evaluation to your deployment process. While the gate is in dry-run mode, the API always returns `pass` and deployments are not impacted by the gate result.
3. After a period of time (for example, 1-2 weeks), check the gate and rule executions on the {{< ui >}}Deployment Gates Evaluations{{< /ui >}} page. The UI shows the real status, so you can see when the gate would have failed and the reason behind it.

> 4. When you are confident that the gate behavior is as you expect, edit the gate and switch the evaluation mode from {{< ui >}}Dry Run{{< /ui >}} to {{< ui >}}Active{{< /ui >}}. Afterwards, the API starts returning the "real" status and deployments start getting promoted or rolled back based on the gate result.

Reviewer: ```suggestion
4. When you are confident that the gate behavior is as you expect, edit the gate and switch the evaluation mode from {{< ui >}}Dry Run{{< /ui >}} to {{< ui >}}Active{{< /ui >}}. Afterwards, the API starts returning the actual status and deployments start getting promoted or rolled back based on the gate result.
```
